From 0cb5e158cd4ec4bf488fcb52999709177a423038 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E2=80=9Chuangyifeng=E2=80=9D?=
 <huangyifeng@eswincomputing.com>
Date: Fri, 31 May 2024 10:22:40 +0800
Subject: [PATCH 068/109] WIN2030-15099:fix(daemon):get power info from SOM
 daemon

Changelogs: 1.Get power info from SOM daemon.
Change-Id: I0d6149071acb5bd70e7fe8431b03afe50b73903f
---
 Core/Inc/hf_common.h  | 10 ++++++++--
 Core/Src/console.c    | 21 ++++++++++-----------
 Core/Src/web-server.c | 34 ++++++++++++++--------------------
 3 files changed, 32 insertions(+), 33 deletions(-)

diff --git a/Core/Inc/hf_common.h b/Core/Inc/hf_common.h
index 35cce30..acc14ee 100644
--- a/Core/Inc/hf_common.h
+++ b/Core/Inc/hf_common.h
@@ -56,6 +56,7 @@ typedef enum {
 	CMD_CONTROL_LED,
 	CMD_PVT_INFO,
 	CMD_BOARD_STATUS,
+	CMD_POWER_INFO,
 	// You can continue adding other command types
 } CommandType;
 
@@ -83,6 +84,12 @@ typedef struct {
 extern QueueHandle_t xUart4MsgQueue;
 extern Message UART4_RxMsg;
 
+typedef struct {
+	uint32_t consumption;
+	uint32_t current;
+	uint32_t voltage;
+} __attribute__((packed)) power_info;
+
 /* The datas(structures) below need to be stored or updated in eeprom */
 // The board info that was initialized(burned) in the eeprom at the factory
 typedef struct {
@@ -117,8 +124,6 @@ typedef struct {
 	uint8_t som_dip_switch_soft_state;	// record the DIP Switch software state of the SOM, it is used for SOM bootsel, bit0---bit3 stand for the DIP0---DIP3
 } SomPwrMgtDIPInfo;
 
-
-
 struct gpio_cmd {
 	uint16_t group;
 	uint16_t pin_num;
@@ -299,6 +304,7 @@ int32_t es_set_rtc_date(struct rtc_date_t *sdate);
 int32_t es_set_rtc_time(struct rtc_time_t *stime);
 int32_t es_get_rtc_date(struct rtc_date_t *sdate);
 int32_t es_get_rtc_time(struct rtc_time_t *stime);
+power_info get_power_info(void);
 
 #ifdef __cplusplus
 }
diff --git a/Core/Src/console.c b/Core/Src/console.c
index 485d63e..c8d3dab 100644
--- a/Core/Src/console.c
+++ b/Core/Src/console.c
@@ -998,17 +998,16 @@ static BaseType_t prvCommandTempGet(char *pcWriteBuffer, size_t xWriteBufferLen,
 */
 static BaseType_t prvCommandPwrDissipationGet(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString)
 {
-    uint32_t millivolt = 0, milliCur = 0, microWatt = 0; //millivolt, milliCurrent, milliwatt
-    int pwrStaus;
-
-    pwrStaus = get_som_power_state();
-    if (pwrStaus == SOM_POWER_ON) {
-        get_board_power(&millivolt, &milliCur, &microWatt);
-    }
-    snprintf(pcWriteBuffer, xWriteBufferLen,"consumption:%ld.%3.3ld(W)  voltage:%ld.%03ld(V)  current:%ld.%03ld(A)\n",
-        microWatt / 1000000, microWatt % 1000000, millivolt / 1000, millivolt % 1000, milliCur / 1000, milliCur % 1000);
-
-    return pdFALSE;
+	uint32_t millivolt = 0, milliCur = 0, microWatt = 0; //millivolt, milliCurrent, milliwatt
+	power_info power_info = {0};
+
+	power_info = get_power_info();
+	millivolt = power_info.voltage;
+	milliCur = power_info.current;
+	microWatt = power_info.consumption;
+	snprintf(pcWriteBuffer, xWriteBufferLen,"consumption:%ld.%3.3ld(W)  voltage:%ld.%03ld(V)  current:%ld.%03ld(A)\n",
+		microWatt / 1000000, microWatt % 1000000, millivolt / 1000, millivolt % 1000, milliCur / 1000, milliCur % 1000);
+	return pdFALSE;
 }
 
 
diff --git a/Core/Src/web-server.c b/Core/Src/web-server.c
index 929cceb..7ed703f 100644
--- a/Core/Src/web-server.c
+++ b/Core/Src/web-server.c
@@ -10484,26 +10484,20 @@ int reset()
 	return ret;
 }
 
+power_info get_power_info(void)
+{
+	power_info power_info = {0};
+	int ret = HAL_OK;
 
-typedef struct  {
-    uint32_t consumption;
-    uint32_t current;
-    uint32_t voltage;
-} POWERInfo;
-
-POWERInfo get_power_info(void){
-
-
-	POWERInfo example = {0};
-
-	if (get_power_status())
-		get_board_power(&example.voltage,&example.current,&example.consumption);
-
-	return example;
+	ret = web_cmd_handle(CMD_POWER_INFO, &power_info, sizeof(power_info), 1000);
+	if (HAL_OK != ret) {
+		web_debug("Failed to get power info %d\n", ret);
+	}
+	web_debug("web call get_power_info, consumption %d, current %d, voltage %d, ret %d\n",
+		power_info->consumption, power_info->current, power_info->voltage, ret);
+	return power_info;
 }
 
-
-
 NETInfo get_net_info(void) {
 	NETInfo example;
 	ip4_addr_t ip4_addr;
@@ -10567,7 +10561,7 @@ int get_pvt_info(PVTInfo *ppvtInfo)
 	int ret = HAL_OK;
 	ret = web_cmd_handle(CMD_PVT_INFO, ppvtInfo, sizeof(PVTInfo), 1000);
 	if (HAL_OK != ret) {
-		web_debug("Faild to get PVT info %d\n", ret);
+		web_debug("Failed to get PVT info %d\n", ret);
 	}
 	web_debug("web call get_pvt_info, cpu_temp %d, npu_temp %d, fan_speed %d, ret %d\n",
 		ppvtInfo->cpu_temp, ppvtInfo->npu_temp,ppvtInfo->fan_speed, ret);
@@ -10727,7 +10721,7 @@ int get_soc_status()
                     found_session_user_name=strdup(found_session->session_data);
                 }
             }
-			
+
         }
 
         char resp_cookies[BUF_SIZE_256] = {0};
@@ -10842,7 +10836,7 @@ int get_soc_status()
                 netconn_write(conn, response_header, strlen(response_header), NETCONN_COPY);
                 netconn_write(conn, json_response, strlen(json_response), NETCONN_COPY);
 			}else if(strcmp(path, "/power_consum")==0 ){
-				POWERInfo power_info=get_power_info();
+				power_info power_info=get_power_info();
 				char json_response[BUF_SIZE_256]={0};
                 char *json_response_patt = "{\"status\":0,\"message\":\"success\",\"data\":{\"consumption\":\"%d\",\"voltage\":\"%d\",\"current\":\"%d\"}}";//0 success, msg
                 sprintf(json_response, json_response_patt,power_info.consumption,power_info.voltage,power_info.current);
-- 
2.25.1

